9.2.26 Lab: Dive into the uhashring repo

In [10]:
from uhashring import HashRing

# create a consistent hash ring of 3 nodes of weight 1
hr = HashRing(nodes=['node1', 'node2', 'node3'])

# get the node name for the 'coconut' key
target_node = hr.get_node('coconut')

print(target_node)

node2


In [12]:
# get the 'coconut' key from the right host's redis instance
hr.get('coconut')

{'hostname': 'node2',
 'instance': None,
 'nodename': 'node2',
 'port': None,
 'vnodes': 160,
 'weight': 1}

3. Where is the hash created? (in the ring.py file)

Line 25 of ring_meta.py

In [13]:
hr.get_node('coconut')

'node2'

In [17]:
hr.get_server('coconut')

(157285054189446768912416291781797853523, 'node2')

4. How many virtual nodes are created for each physical node?

In [18]:
hr.get('coconut')

{'hostname': 'node2',
 'instance': None,
 'nodename': 'node2',
 'port': None,
 'vnodes': 160,
 'weight': 1}

160 (see ring.py line 30)

Where are virtual nodes created?

What is the naming convention for virtual nodes?

Ketama-style libraries (and uhashring, following that convention) name each vnode as "{nodename}-{index}", e.g. node1-0, node1-1, ... node1-159. That string is what actually gets hashed to produce each vnode's ring position

7. Return all the ring positions for node 1.

In [23]:
hr.runtime._node_points['node1']

[81200024130818730098147654587186424104,
 20846744489285949295898190393411514004,
 246086855014282440052137811901676469602,
 246117650072666980773543863665245303742,
 162075670587272527109963292863522227373,
 128653884373862206501156681709473025385,
 509571543463130690647202895977041641,
 117736815197465365305039462959815752314,
 112303069909446470524039559147856547866,
 109407622066901525696465599565309600177,
 314883097326498975757221000650465148776,
 183817067659701951718157235419476193102,
 127486792529685998817100610489168692462,
 302580115757318666714422081099847858639,
 202729871933705400049611271005266041358,
 139834255818146452654516042222077075762,
 136147817149905781122963680777324820289,
 64401656268765730759598918999342764002,
 162943191358880783835618505974839352700,
 281997554278523799200601154854544312551,
 96211305920231675920439599381435002027,
 154687340403087519794218438121580327058,
 278919706586509500439137953377675074335,
 140805955665183275075443969460185746534,

Revisit the example from Basic usage and look at the target_n for the key 'coconut'.